sendbuffer.go

```diff
@@ -100,17 +104,41 @@ func (buf *sendBuffer) sendLoop(w io.Writer) {
 	}
 }
 
+func (buf *sendBuffer) send(b []byte, writeDeadline time.Time) (int, error) {
+	buf.muClosing.RLock()
+	defer buf.muClosing.RUnlock()
```

Comment:
Something that I don't love about this is that it adds a mutex to the critical path of sending (which next to receive is the most frequent op). The `defer` also adds a little overhead. Is there a way we could do this without a lock?

Reply:
Yeah, I don't like it either. I tried a few alternatives before choosing Mutex, but we have to sequentialize the access to `buf.in` to avoid sending on closed channel. IIRC sending to channel is even more expensive than mutex locking and defer, so maybe it's acceptable?

Reply:
With this bench file https://play.golang.org/p/hUzJfbaMSuH I got below result. So mutex + defer is comparable to channel. Eliminating `defer` is straightforward. I'll do that.
```
pkg: github.com/getlantern/lampshade/a
BenchmarkChannel-4      	20000000	        71.3 ns/op
BenchmarkMutex-4        	100000000	        19.2 ns/op
BenchmarkMutexDefer-4   	20000000	        63.4 ns/op
```